File: scripts/artifact_registry.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["pyyaml"]
# ///
"""Registry-backed artifact identity projection for Agentera scripts."""

from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml


REPO_ROOT = Path(__file__).resolve().parent.parent
ARTIFACT_SCHEMAS_DIR = REPO_ROOT / "skills" / "agentera" / "schemas" / "artifacts"
REGISTRY_MODEL_PATH = REPO_ROOT / "references" / "artifacts" / "artifact-registry-interface-model.yaml"


@dataclass(frozen=True)
class ArtifactRecord:
    artifact_id: str
    display_name: str
    default_path: str
    producers: set[str]
    consumers: set[str]
    artifact_type: str
    scope: str
    path_template: dict[str, Any] | None
    docs_yaml_can_override_path: bool


def load_yaml(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as fh:
        data = yaml.safe_load(fh)
    return data if isinstance(data, dict) else {}


def as_set(value: Any) -> set[str]:
    if value is None:
        return set()
    if isinstance(value, str):
        return {value}
    if isinstance(value, list):
        return {str(v) for v in value}
    return {str(value)}


def normalize_path(path: str) -> str:
    """Normalize persisted prose-era variants before comparison/resolution."""
    p = str(path).strip()
    p = re.sub(r"\s*\([^)]*\)\s*$", "", p)
    p = re.sub(r"\s+or\s+mapped\s+path\s+per\s+(?:docs\.yaml|DOCS\.md)$", "", p)
    p = p.replace("<objective-name>", "<name>")
    return p.strip()


def _schema_metas(artifact_schemas_dir: Path) -> dict[str, dict[str, Any]]:
    metas: dict[str, dict[str, Any]] = {}
    if not artifact_schemas_dir.is_dir():
        return metas
    for schema_path in sorted(artifact_schemas_dir.glob("*.yaml")):
        meta = load_yaml(schema_path).get("meta")
        if not isinstance(meta, dict):
            continue
        artifact_id = str(meta.get("name", "")).strip()
        if artifact_id:
            metas[artifact_id] = meta
    return metas


def load_artifact_registry(
    artifact_schemas_dir: Path = ARTIFACT_SCHEMAS_DIR,
    registry_model_path: Path = REGISTRY_MODEL_PATH,
) -> dict[str, ArtifactRecord]:
    model = load_yaml(registry_model_path)
    metas = _schema_metas(artifact_schemas_dir)
    records: dict[str, ArtifactRecord] = {}

    for scope, identities in model.get("required_artifact_identities", {}).items():
        if not isinstance(identities, list):
            continue
        for identity in identities:
            if not isinstance(identity, dict):
                continue
            artifact_id = str(identity.get("artifact_id", "")).strip()
            meta = metas.get(artifact_id)
            if not artifact_id or not meta:
                continue
            records[artifact_id] = ArtifactRecord(
                artifact_id=artifact_id,
                display_name=str(identity.get("display_name", "")).strip(),
                default_path=normalize_path(str(identity.get("default_path", meta.get("path", "")))),
                producers=as_set(meta.get("producer")),
                consumers=as_set(meta.get("consumers")),
                artifact_type=str(meta.get("artifact_type", "")).strip(),
                scope=str(scope),
                path_template=identity.get("path_template") if isinstance(identity.get("path_template"), dict) else None,
                docs_yaml_can_override_path=True,
            )

    for special in model.get("explicit_special_cases", []):
        if not isinstance(special, dict):
            continue
        artifact_id = str(special.get("artifact_id", "")).strip()
        if not artifact_id:
            continue
        records[artifact_id] = ArtifactRecord(
            artifact_id=artifact_id,
            display_name=str(special.get("display_name", "")).strip(),
            default_path=normalize_path(str(special.get("default_path", ""))),
            producers=as_set(special.get("producers")),
            consumers=as_set(special.get("consumers")),
            artifact_type=str(special.get("artifact_type", "")).strip(),
            scope=str(special.get("scope", "")).strip(),
            path_template=special.get("path_template") if isinstance(special.get("path_template"), dict) else None,
            docs_yaml_can_override_path=bool(special.get("docs_yaml_can_override_path")),
        )

    return records


def load_docs_path_overrides(project_root: Path) -> dict[str, str]:
    docs_path = project_root / ".agentera" / "docs.yaml"
    if not docs_path.exists():
        return {}
    try:
        data = yaml.safe_load(docs_path.read_text(encoding="utf-8"))
    except Exception:
        return {}
    mapping = data.get("mapping") if isinstance(data, dict) else None
    if not isinstance(mapping, list):
        return {}
    overrides: dict[str, str] = {}
    for entry in mapping:
        if not isinstance(entry, dict):
            continue
        artifact = entry.get("artifact")
        path = entry.get("path")
        if isinstance(artifact, str) and isinstance(path, str):
            overrides[artifact] = path
    return overrides


def resolve_artifact_path(
    record: ArtifactRecord,
    project_root: Path,
    active_objective_name: str | None = None,
) -> Path:
    artifact_path = record.default_path
    overrides = load_docs_path_overrides(project_root)
    if record.docs_yaml_can_override_path and record.display_name in overrides:
        artifact_path = overrides[record.display_name]
    if "<name>" in artifact_path and active_objective_name:
        artifact_path = artifact_path.replace("<name>", active_objective_name)
    if artifact_path.startswith("$PROFILERA_PROFILE_DIR/"):
        explicit = os.environ.get("PROFILERA_PROFILE_DIR")
        suffix = artifact_path.removeprefix("$PROFILERA_PROFILE_DIR/")
        if explicit:
            return Path(explicit) / suffix
        xdg = os.environ.get("XDG_DATA_HOME")
        base = Path(xdg) if xdg else Path.home() / ".local" / "share"
        return base / "agentera" / suffix
    path = Path(artifact_path)
    if path.is_absolute():
        return path
    return project_root / path

File: packages/cli/src/registries/artifactRegistry.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import { loadYamlMapping } from "../core/yaml.js";
import { resolveProfileDirOverride } from "../core/envPaths.js";
import { resolvePath } from "../core/paths.js";
import { resolveSourceRoot } from "../core/sourceRoot.js";
import { resolveCandidate } from "../state/installRoot.js";

/**
 * Registry-backed artifact identity projection. Faithful TS port of
 * `scripts/artifact_registry.py`.
 */

export interface ArtifactRecord {
  artifactId: string;
  displayName: string;
  defaultPath: string;
  producers: Set<string>;
  consumers: Set<string>;
  artifactType: string;
  scope: string;
  pathTemplate: Record<string, unknown> | null;
  docsYamlCanOverridePath: boolean;
}

function sourceRoot(): string {
  return resolveSourceRoot();
}

export function artifactSchemasDir(root: string = sourceRoot()): string {
  return path.join(root, "skills", "agentera", "schemas", "artifacts");
}

export function registryModelPath(root: string = sourceRoot()): string {
  return path.join(root, "references", "artifacts", "artifact-registry-interface-model.yaml");
}

function loadYaml(p: string): Record<string, unknown> {
  return loadYamlMapping(fs.readFileSync(p, "utf8"));
}

export function asSet(value: unknown): Set<string> {
  if (value === null || value === undefined) {
    return new Set();
  }
  if (typeof value === "string") {
    return new Set([value]);
  }
  if (Array.isArray(value)) {
    return new Set(value.map((v) => String(v)));
  }
  return new Set([String(value)]);
}

export function normalizePath(p: string): string {
  let s = String(p).trim();
  s = s.replace(/\s*\([^)]*\)\s*$/, "");
  s = s.replace(/\s+or\s+mapped\s+path\s+per\s+(?:docs\.yaml|DOCS\.md)$/, "");
  s = s.replace(/<objective-name>/g, "<name>");
  return s.trim();
}

const URI_SCHEME_RE = /^[A-Za-z][A-Za-z0-9+.-]*:/;
const ENCODED_TRAVERSAL_RE = /%(?:2e|2f|5c)/i;
const WINDOWS_DRIVE_RE = /^[A-Za-z]:[\\/]/;

function rejectUnsafeArtifactPath(p: string, artifactId: string): void {
  if ([...p].some((char) => char.charCodeAt(0) <= 31 || char.charCodeAt(0) === 127)) {
    throw new Error(`artifact '${artifactId}' path contains control characters`);
  }
  if (ENCODED_TRAVERSAL_RE.test(p)) {
    throw new Error(`artifact '${artifactId}' path contains encoded traversal or separators`);
  }
  if (URI_SCHEME_RE.test(p) && !WINDOWS_DRIVE_RE.test(p)) {
    throw new Error(`artifact '${artifactId}' path uses unsupported URI syntax`);
  }
  const parts = p.split(/[\\/]/);
  if (parts.some((part) => part === "..")) {
    throw new Error(`artifact '${artifactId}' path contains traversal segments`);
  }
}

function projectPath(projectRoot: string, artifactPath: string, artifactId: string): string {
  rejectUnsafeArtifactPath(artifactPath, artifactId);
  const resolvedProject = resolvePath(projectRoot);
  const resolved = path.isAbsolute(artifactPath)
    ? resolvePath(artifactPath)
    : resolvePath(path.join(resolvedProject, artifactPath));
  const rel = path.relative(resolvedProject, resolved);
  if (rel !== "" && (rel.startsWith("..") || path.isAbsolute(rel))) {
    throw new Error(`artifact '${artifactId}' path escapes the project boundary`);
  }
  return resolved;
}

export const EXPECTED_ARTIFACT_SCHEMA_VERSION = "1.0.0";

function readSchemaMeta(file: string): Record<string, unknown> | null {
  const meta = loadYaml(file).meta;
  if (!meta || typeof meta !== "object" || Array.isArray(meta)) {
    return null;
  }
  return meta as Record<string, unknown>;
}

function warnMalformedSchemaMeta(name: string): void {
  process.stderr.write(
    `warning: artifact schema ${name} is missing meta; expected v3 schema version ${EXPECTED_ARTIFACT_SCHEMA_VERSION}\n`,
  );
}

function warnSchemaVersionMismatch(meta: Record<string, unknown>, name: string): void {
  const version = String(meta.version ?? "").trim();
  if (!version || version !== EXPECTED_ARTIFACT_SCHEMA_VERSION) {
    const reported = version || "(missing)";
    process.stderr.write(
      `warning: artifact schema ${name} version ${reported} does not match expected v3 schema version ${EXPECTED_ARTIFACT_SCHEMA_VERSION}\n`,
    );
  }
}

function listSchemaFiles(dir: string): string[] {
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) {
    return [];
  }
  return fs
    .readdirSync(dir)
    .filter((name) => name.endsWith(".yaml"))
    .sort();
}

/**
 * Full schema-meta index over an artifact schemas directory. Emits malformed-meta
 * and version-mismatch warnings for every schema file; the full loader relies on
 * this strict coverage to surface drift across the whole identity set.
 */
function schemaMetas(dir: string): Map<string, Record<string, unknown>> {
  const metas = new Map<string, Record<string, unknown>>();
  for (const name of listSchemaFiles(dir)) {
    const meta = readSchemaMeta(path.join(dir, name));
    if (!meta) {
      warnMalformedSchemaMeta(name);
      continue;
    }
    warnSchemaVersionMismatch(meta, name);
    const artifactId = String(meta.name ?? "").trim();
    if (artifactId) {
      metas.set(artifactId, meta);
    }
  }
  return metas;
}

/**
 * Targeted schema-meta lookup for a single artifact identity. Walks the schemas
 * directory only to find the file whose `meta.name` matches `artifactId`, and
 * emits warnings only for that schema. Unrelated malformed or version-mismatched
 * schemas in the same directory stay silent so callers resolving a single
 * identity (e.g. `profile`) don't fan out noise across the required-identity set.
 */
function loadSchemaMetaForArtifact(
  dir: string,
  artifactId: string,
): Record<string, unknown> | null {
  for (const name of listSchemaFiles(dir)) {
    let meta: Record<string, unknown> | null;
    try {
      meta = readSchemaMeta(path.join(dir, name));
    } catch {
      continue;
    }
    if (!meta) continue;
    if (String(meta.name ?? "").trim() !== artifactId) continue;
    warnSchemaVersionMismatch(meta, name);
    return meta;
  }
  return null;
}

function findRequiredIdentity(
  model: Record<string, unknown>,
  artifactId: string,
): { scope: string; identity: Record<string, unknown> } | null {
  const identities = (model.required_artifact_identities ?? {}) as Record<string, unknown>;
  for (const [scope, identityList] of Object.entries(identities)) {
    if (!Array.isArray(identityList)) continue;
    for (const identity of identityList) {
      if (!identity || typeof identity !== "object" || Array.isArray(identity)) continue;
      const id = identity as Record<string, unknown>;
      if (String(id.artifact_id ?? "").trim() === artifactId) {
        return { scope, identity: id };
      }
    }
  }
  return null;
}

function findSpecialCase(
  model: Record<string, unknown>,
  artifactId: string,
): Record<string, unknown> | null {
  const specialCases = (model.explicit_special_cases ?? []) as unknown[];
  for (const special of specialCases) {
    if (!special || typeof special !== "object" || Array.isArray(special)) continue;
    const sp = special as Record<string, unknown>;
    if (String(sp.artifact_id ?? "").trim() === artifactId) {
      return sp;
    }
  }
  return null;
}

function buildRequiredRecord(
  scope: string,
  identity: Record<string, unknown>,
  meta: Record<string, unknown>,
): ArtifactRecord {
  const template = identity.path_template;
  return {
    artifactId: String(identity.artifact_id ?? "").trim(),
    displayName: String(identity.display_name ?? "").trim(),
    defaultPath: normalizePath(String(identity.default_path ?? meta.path ?? "")),
    producers: asSet(meta.producer),
    consumers: asSet(meta.consumers),
    artifactType: String(meta.artifact_type ?? "").trim(),
    scope: String(scope),
    pathTemplate:
      template && typeof template === "object" && !Array.isArray(template)
        ? (template as Record<string, unknown>)
        : null,
    docsYamlCanOverridePath: true,
  };
}

function buildSpecialCaseRecord(sp: Record<string, unknown>): ArtifactRecord {
  const template = sp.path_template;
  return {
    artifactId: String(sp.artifact_id ?? "").trim(),
    displayName: String(sp.display_name ?? "").trim(),
    defaultPath: normalizePath(String(sp.default_path ?? "")),
    producers: asSet(sp.producers),
    consumers: asSet(sp.consumers),
    artifactType: String(sp.artifact_type ?? "").trim(),
    scope: String(sp.scope ?? "").trim(),
    pathTemplate:
      template && typeof template === "object" && !Array.isArray(template)
        ? (template as Record<string, unknown>)
        : null,
    docsYamlCanOverridePath: Boolean(sp.docs_yaml_can_override_path),
  };
}

export function loadArtifactRegistry(
  artifactSchemasDirPath: string = artifactSchemasDir(),
  registryModelPathArg: string = registryModelPath(),
): Map<string, ArtifactRecord> {
  const model = loadYaml(registryModelPathArg);
  const metas = schemaMetas(artifactSchemasDirPath);
  const records = new Map<string, ArtifactRecord>();

  const identities = (model.required_artifact_identities ?? {}) as Record<string, unknown>;
  for (const [scope, identityList] of Object.entries(identities)) {
    if (!Array.isArray(identityList)) {
      continue;
    }
    for (const identity of identityList) {
      if (!identity || typeof identity !== "object" || Array.isArray(identity)) {
        continue;
      }
      const id = identity as Record<string, unknown>;
      const artifactId = String(id.artifact_id ?? "").trim();
      if (!artifactId) {
        continue;
      }
      const meta = metas.get(artifactId);
      if (!meta) {
        process.stderr.write(
          `warning: required artifact identity '${artifactId}' has no matching schema file in ${artifactSchemasDirPath}\n`,
        );
        continue;
      }
      records.set(artifactId, buildRequiredRecord(scope, id, meta));
    }
  }

  const specialCases = (model.explicit_special_cases ?? []) as unknown[];
  for (const special of specialCases) {
    if (!special || typeof special !== "object" || Array.isArray(special)) {
      continue;
    }
    const sp = special as Record<string, unknown>;
    const artifactId = String(sp.artifact_id ?? "").trim();
    if (!artifactId) {
      continue;
    }
    records.set(artifactId, buildSpecialCaseRecord(sp));
  }

  return records;
}

/**
 * Targeted lookup of a single artifact record. Resolves against the same
 * model/schema authority as `loadArtifactRegistry` but inspects only the
 * requested identity, so a special case such as `profile` loads without
 * walking or warning about unrelated required schemas. A requested required
 * identity still warns and returns `undefined` when its matching schema file is
 * missing from the schemas directory.
 */
export function loadArtifactRecord(
  artifactId: string,
  artifactSchemasDirPath: string = artifactSchemasDir(),
  registryModelPathArg: string = registryModelPath(),
): ArtifactRecord | undefined {
  const model = loadYaml(registryModelPathArg);
  const specialCase = findSpecialCase(model, artifactId);
  if (specialCase) {
    return buildSpecialCaseRecord(specialCase);
  }
  const required = findRequiredIdentity(model, artifactId);
  if (!required) {
    return undefined;
  }
  const meta = loadSchemaMetaForArtifact(artifactSchemasDirPath, artifactId);
  if (!meta) {
    process.stderr.write(
      `warning: required artifact identity '${artifactId}' has no matching schema file in ${artifactSchemasDirPath}\n`,
    );
    return undefined;
  }
  return buildRequiredRecord(required.scope, required.identity, meta);
}

export function docsPathOverridesFromBytes(bytes: string | Buffer, strict = false): Record<string, string> {
  let data: Record<string, unknown>;
  try {
    data = loadYamlMapping(bytes.toString());
  } catch (exc) {
    if (strict) throw new Error(`failed to load docs path overrides: ${(exc as Error).message}`);
    process.stderr.write(`warning: failed to load docs path overrides: ${(exc as Error).message}\n`);
    return {};
  }
  const mapping = data.mapping;
  if (!Array.isArray(mapping)) {
    if (strict && mapping !== undefined) throw new Error("failed to load docs path overrides: mapping must be a list");
    return {};
  }
  const overrides: Record<string, string> = {};
  for (const entry of mapping) {
    if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
      if (strict) throw new Error("failed to load docs path overrides: every mapping entry must be a mapping");
      continue;
    }
    const e = entry as Record<string, unknown>;
    if (typeof e.artifact === "string" && typeof e.path === "string") overrides[e.artifact] = e.path;
    else if (strict) throw new Error("failed to load docs path overrides: every mapping entry requires string artifact and path fields");
  }
  return overrides;
}

export function loadDocsPathOverrides(projectRoot: string, strict = false): Record<string, string> {
  const docsPath = path.join(projectRoot, ".agentera", "docs.yaml");
  if (!fs.existsSync(docsPath)) {
    return {};
  }
  let bytes: Buffer;
  try {
    bytes = fs.readFileSync(docsPath);
  } catch (exc) {
    if (strict) throw new Error(`failed to load docs path overrides: ${(exc as Error).message}`);
    process.stderr.write(
      `warning: failed to load docs path overrides: ${(exc as Error).message}\n`,
    );
    return {};
  }
  return docsPathOverridesFromBytes(bytes, strict);
}

export interface ResolveArtifactPathOptions {
  activeObjectiveName?: string | null;
  strictWrite?: boolean;
  docsPathOverrides?: Readonly<Record<string, string>>;
}

function nearestExistingAncestor(p: string): string {
  let current = p;
  while (!fs.existsSync(current)) {
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return fs.realpathSync.native(current);
}

export function assertRealpathBoundary(
  projectRoot: string,
  artifactPath: string,
  artifactId: string,
): void {
  const projectReal = nearestExistingAncestor(resolvePath(projectRoot));
  const ancestorReal = nearestExistingAncestor(artifactPath);
  const rel = path.relative(projectReal, ancestorReal);
  if (rel !== "" && (rel.startsWith("..") || path.isAbsolute(rel))) {
    throw new Error(`artifact '${artifactId}' path escapes the project boundary`);
  }
}

export function resolveArtifactPath(
  record: ArtifactRecord,
  projectRoot: string,
  activeObjectiveNameOrOptions: string | null | ResolveArtifactPathOptions = null,
  env: Record<string, string | undefined> = process.env,
): string {
  const options =
    typeof activeObjectiveNameOrOptions === "object" && activeObjectiveNameOrOptions !== null
      ? activeObjectiveNameOrOptions
      : { activeObjectiveName: activeObjectiveNameOrOptions };
  const activeObjectiveName = options.activeObjectiveName ?? null;
  let artifactPath = record.defaultPath;
  const overrides = options.docsPathOverrides ?? loadDocsPathOverrides(projectRoot, options.strictWrite === true);
  if (record.docsYamlCanOverridePath && record.displayName in overrides) {
    artifactPath = overrides[record.displayName];
  }
  if (artifactPath.includes("<name>") && activeObjectiveName) {
    artifactPath = artifactPath.replace(/<name>/g, activeObjectiveName);
  }
  const profileDirPrefixes = ["$AGENTERA_PROFILE_DIR/", "$PROFILERA_PROFILE_DIR/"] as const;
  for (const prefix of profileDirPrefixes) {
    if (artifactPath.startsWith(prefix)) {
      const suffix = artifactPath.slice(prefix.length);
      const explicit = resolveProfileDirOverride(env);
      if (explicit) {
        return path.join(explicit, suffix);
      }
      const [base] = resolveCandidate(null, { env, home: os.homedir() });
      return path.join(base, suffix);
    }
  }
  const resolved = projectPath(projectRoot, artifactPath, record.artifactId);
  if (options.strictWrite) assertRealpathBoundary(projectRoot, resolved, record.artifactId);
  return resolved;
}
